// react-native-barcode-scanners/src/Scanner.js
'use strict';
import React, { Component } from 'react';
import {
  StyleSheet,
  View,
} from 'react-native';
import { RNCamera } from 'react-native-camera';
import PropTypes from 'prop-types';

import ScannerMask from './ScannerMask';
import ScannerFooter from './ScannerFooter';

export default class Scanner extends Component {

  static propTypes = {
    Title: PropTypes.string, // 顶部栏的名称 default {Title}
    isEnableTopBar: PropTypes.boolean, // 是否启用顶部栏目 default {true}
    RenderTopBar: PropTypes.element, // 顶部栏自定义组件
    topBarBackgroundColor: PropTypes.string, // 顶部栏背景色 default {rgba(0,0,0,0.1)}

    isEnableAnimationScan: PropTypes.boolean, //是否启用扫描动画 default {true}
    isEnableMask: PropTypes.boolean, //是否启用背景蒙版 default {true}
    isEnableScanBorder: PropTypes.boolean, // 是否启用扫描的边框 default{true}
    isEnableDiscernPicture: PropTypes.boolean, // 是否启用从相册获取识别图片二维码 {true}
    RenderScanLine: PropTypes.element, // 自定义扫描组件
    maskColor: PropTypes.string, // 蒙版的背景色 default {rgba(0,0,0,0.4)}
    scanAngleColor: PropTypes.string, // 扫描角的背景色 {rgba(250,250,250,0.4)}
    scanAngleWidth: PropTypes.number, // 扫描角的长度 {28}
    scanAngleHeight: PropTypes.number, // 扫描角的高度 {5}
    scanBoxSize: PropTypes.number, // 扫描框的大小 {WINDOW_WIDTH * 0.6}
    scanBorderColor: PropTypes.boolean, // 边框的颜色 default{ rgba(250,250,250,0.3) }
    hintScanText: PropTypes.string, // 扫描提示文字 default {Put the barcode into the box}
    hintOrText: PropTypes.string, // 提示文字其他选择 default {OR}
    hintPictureGalleryText: PropTypes.string, // 从相册选取二维码 default {Choose barcode from album}
    hintOpenLightText: PropTypes.string, // 提示文字打开手电筒 default{ON}
    hintCloseLightText: PropTypes.string, // 提示文字关闭手电筒 default{OFF}
    lightBackgroundColor: PropTypes.string, // 手电筒脚步位置的背景色 default{#181D20}

    onBack: PropTypes.func, // 顶部栏返回点击
    onBarCodeRead: PropTypes.func, // 识别到条码
    onBarCodeReadByGalleryStart: PropTypes.func, // 当开始识别相册图片的二维码
    onBarCodeReadByGalleryFailure: PropTypes.func, // 没有识别到图片中的二维码
  }
  constructor(props){
    super(props)
    this.state = {
      isOpenLight: false,
      isEnableCameraScan: true,
      scanInterval: 1500
    }
  }
  componentDidMount () {}
  render() {
    const props = this.props
    return (
      <View style={styles.container}>
        <RNCamera
          ref={ref => {
            this.camera = ref;
          }}
          style={styles.preview}
          onBarCodeRead={this.state.isEnableCameraScan === true ? (code)=>this.onBarCodeReadByCamera.bind(this)(code) : null}
          type={RNCamera.Constants.Type.back}
          flashMode={this.state.isOpenLight === true ? RNCamera.Constants.FlashMode.torch : RNCamera.Constants.FlashMode.off}
          permissionDialogTitle={'Permission to use camera'}
          permissionDialogMessage={'We need your permission to use your camera phone'}
        >
          <ScannerMask
            Title={props.Title}
            isEnableAnimationScan={props.isEnableAnimationScan}
            isEnableTopBar={props.isEnableTopBar}
            RenderScanLine={props.RenderScanLine}
            RenderTopBar={props.RenderTopBar}
            topBarBackgroundColor={props.topBarBackgroundColor}
            isEnableMask={props.isEnableMask}
            isEnableScanBorder={props.isEnableScanBorder}
            isEnableDiscernPicture={props.isEnableDiscernPicture}
            maskColor={props.maskColor}
            scanAngleColor={props.scanAngleColor}
            scanAngleWidth={props.scanAngleWidth}
            scanAngleHeight={props.scanAngleHeight}
            scanBoxSize={props.scanBoxSize}
            scanBorderColor={props.scanBorderColor}
            hintScanText={props.hintScanText}
            hintOrText={props.hintOrText}
            hintPictureGalleryText={props.hintPictureGalleryText}

            onBack={()=>props.onBack && props.onBack()}
            onBarCodeReadByGalleryStart={()=>props.onBarCodeReadByGalleryStart && props.onBarCodeReadByGalleryStart()}
            onBarCodeReadByGalleryFailure={()=>props.onBarCodeReadByGalleryFailure && props.onBarCodeReadByGalleryFailure()}
            onBarCodeRead={(data)=>{this.onBarCodeRead.bind(this)(data)}}
          />
        </RNCamera>
        <ScannerFooter
          hintOpenLightText={props.hintOpenLightText}
          hintCloseLightText={props.hintCloseLightText}
          lightBackgroundColor={props.lightBackgroundColor}
          isOpenLight={this.state.isOpenLight}
          onOpenLight={()=>this.handlerLight.bind(this)(true)}
          onCloseLight={()=>this.handlerLight.bind(this)(false)}
        />
      </View>
    );
  }
  // 相机识别到条码
  onBarCodeReadByCamera(data){
    let _this = this
    _this.onBarCodeRead.bind(this)(data)
    _this.setState({
      isEnableCameraScan: false
    }, ()=>{
      _this.timer && clearTimeout(_this.timer);
      _this.timer = setTimeout(()=>{
        _this.setState({
          isEnableCameraScan: true
        })
      }, _this.scanInterval)
    })
  }
  componentWillUnmount() {
    this.timer && clearTimeout(this.timer);
  }

  // 识别到条码
  onBarCodeRead(data){
    this.handlerLight.bind(this)(false)
    this.props.onBarCodeRead && this.props.onBarCodeRead(data)
  }
  // 控制开灯关灯
  handlerLight(stutas){
    this.setState({
      isOpenLight: stutas
    })
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
  },
  preview: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center'
  }
})
