// react-native-barcode-scanners/src/ScannerFooter.js
'use strict';
import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  Image
} from 'react-native';

import PropTypes from 'prop-types';

export default class ScannerFooter extends Component {
  static propTypes = {
    onCloseLight: PropTypes.func, // 当关闭手电筒事件
    onOpenLight: PropTypes.func, // 当打开手电筒事件
    isOpenLight: PropTypes.boolean, // 是否打开手电筒
    hintOpenLightText: PropTypes.string, // 提示文字打开手电筒 default{ON}
    hintCloseLightText: PropTypes.string, // 提示文字关闭手电筒 default{OFF}
    lightBackgroundColor: PropTypes.string, // 手电筒脚步位置的背景色 default{}
  };
  constructor(props){
    super(props)
  }
  render() {
    const props = this.props
    const styles = this._getStyles.call(this)
    let hintOpenLightText = props.hintOpenLightText ? props.hintOpenLightText : 'ON'
    let hintCloseLightText = props.hintCloseLightText ? props.hintCloseLightText : 'OFF'
    return (
      <View style={styles.footer}>
        <TouchableOpacity
          onPress={()=>this.handlerLight.bind(this)()}
        >
          <View style={{justifyContent: 'center', alignItems: 'center'}}>
            <View style={{backgroundColor: '#ffffff', alignItems: 'center', justifyContent: 'center', padding: 20, borderRadius: 54, width: 54, height: 54}}>
              {props.isOpenLight === true ?
                <Image style={{height: 28, width: 28}} source={require('./image/light-bulb-open.png')}/>
                :
                <Image style={{height: 28, width: 28}} source={require('./image/light-bulb-close.png')}/>
              }
            </View>
            <Text  style={{fontSize: 16, marginTop: 10, textAlign: 'center', color: '#f5f5f5'}}>
              {props.isOpenLight === true ? hintCloseLightText : hintOpenLightText}
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
  // 手电筒操作
  handlerLight () {
    let status = this.props.isOpenLight
    if (status === true) { // 开启状态，执行关闭
      this.props.onCloseLight && this.props.onCloseLight()
    } else {
      this.props.onOpenLight && this.props.onOpenLight()
    }
  }
  _getStyles(){
    return StyleSheet.create({
      footer: {
        paddingTop: 24,
        paddingBottom: 24,
        backgroundColor: this.props.lightBackgroundColor !=null ? this.props.lightBackgroundColor : '#181D20',
        justifyContent: 'center',
        alignItems: 'center',
      }
    })
  }
}
